feat(browser): npm nerdgraph tutorial (#12891)
newrelic/docs-website: src/content/docs/browser/browser-monitoring/tutorials/combining-npm-nerdgraph.mdx

---

> title: Combinging NPM package and NerdGraph

Reviewer: Should it be NPM

// The agent loader code executes immediately on instantiation.
new BrowserAgent()

// Remaining code
```
</Step>

<Step>
### Modify the graphql query

Modify the graphql query to retrieve the browser application config as a JSON encoded script string. Run the new query remembering to replace `<guid>` with the browser application entity guid retrieved earlier.

```graphql
query {
  actor {
    entity(guid: '<guid>') {
      ... on BrowserApplicationEntity {
        guid
        name
        browserProperties {
          jsConfigScript
        }
      }
    }
  }
}
```
</Step>

<Step>
### Install the config

Copy the contents of the `jsConfigScript` property. You will need to run the results through a JSON parser or replace the values of `\"` with `"`. The result will be a string containing a `<script>` element with all the browser application config information. Place the `<script>` element at the top of the `<head>` element, but below any `<meta>` elements, of your website HTML.
</Step>
</Steps>

## Taking it further

This tutorial only scratches the surface of how you might use the NerdGraph APIs and the browser agent npm package. Some ideas for further integration:

> - Integrate the browser agent into your website builder application and use the NerdGraph APIs to retrieve the browser application config for each of your customers when they enable New Relic monitoring within your management dashboard.

Reviewer 2: ```suggestion
- Integrate the browser agent into your SaaS platform using the npm package and use the NerdGraph APIs to retrieve browser application configurations specific to each of your customers as they enable New Relic monitoring within your management dashboard.
```

Reviewer 2: I've got nothing against website builders, but I thought maybe we could cast a broader net here.

> - Create a shared browser agent component in your organization's component library and automatically injects the configuration into the HTML using an CDN edge function that calls the NerdGraph API.

Reviewer 2: ```suggestion
- Create a generic shared browser agent component in your organization's component library using the npm package, then automatically inject a configuration block into the HTML using a CDN edge function that calls the NerdGraph API.
```

Reviewer 2: I'm not that experienced with edge functions, but I think the implication here is that the edge function would be responsible for determining which config block to use based on where the app is served, which application the component is part of, or other relevant variables. Can an edge function be deterministic and cache output based on inputs? Otherwise might be a little spammy on the NerdGraph queries. I sense that some variation of what you're describing could be very cool. Just curious how it could work.

Author: Depends on the service and architecture. Akamai has EdgeKV which can be used to store the NG response and recall it.

> - Modify your CI/CD pipeline to populate the source code with different browser application configurations from the NerdGraph API based on what environment is being deployed to.

Reviewer 2: ```suggestion
- Call the NerdGraph API in your CI/CD pipeline to populate your source code with a different browser agent configuration for each of your deployment environments.
```
